README.rst
.. -*- coding: utf-8 -*-
.. :Project:   metapensiero.asyncio.tasklocal -- An asyncio's Task-local variable container
.. :Created:   dom 09 ago 2015 12:57:35 CEST
.. :Author:    Alberto Berti <alberto@metapensiero.it>
.. :License:   GNU General Public License version 3 or later
.. :Copyright: Copyright (C) 2015 Alberto Berti
..

================================
 metapensiero.asyncio.tasklocal
================================

:author: Alberto Berti
:contact: alberto@metapensiero.it
:license: GNU General Public License version 3 or later

An asyncio's Task-local variable container
==========================================

Installation
++++++++++++

To install the package execute the following command::

  $ pip install metapensiero.asyncio.tasklocal

Usage
+++++

A drop-in replacement for ``threading.local``  object that stores
per-task variables (instead of per-thread). You can use it like this:

.. code:: python

  import asyncio
  from metapensiero.asyncio.tasklocal import Local

  mylocal = Local()

  @asyncio.coroutine
  def task_a():
      mylocal.a_var = 'foo'
      yield from asyncio.sleep(0.001) # switches to task_b
      assert mylocal.a_var == 'foo'
      print(mylocal.a_var)

  @asyncio.coroutine
  def task_b():
      mylocal.a_var = 'bar'
      yield from asyncio.sleep(0.001) # switches back to task_a
      assert mylocal.a_var == 'bar'
      print(mylocal.a_var)

  loop = asyncio.get_event_loop()
  a = asyncio.Task(task_a())
  b = asyncio.Task(task_b())
  loop.run_until_complete(asyncio.wait((a, b)))

  # this prints out:
  # foo
  # bar

Really the kind of object to be tracked as "current" is configurable
by passing-in a different ``Discriminator`` instance to the ``Local``
class.

By default it will use the ``TaskDiscriminator`` supplied with the
package. It will use standard ``asyncio``'s loop detection techniques
(by delegating it to ``asyncio.Task.current_task()``) but if you want
to track tasks of a particular event loop, you can supply your own
``TaskDiscriminator`` instance to the ``Local`` instance
initialization, like this:

.. code:: python

  from metapensiero.asyncio.tasklocal import Local, TaskDiscriminator

  mylocal = Local(discriminator=TaskDiscriminator(loop=my_loop))

You can also track another set of object by subclassing the
``BaseDiscriminator`` class:

.. code:: python

  from metapensiero.asyncio.tasklocal import BaseDiscriminator

  class ASampleSessionDiscriminator(BaseDiscriminator):

      def current(self):
          return current_session()

  mylocal = Local(discriminator=ASampleDiscriminator())

You have to implement at least the ``current`` method.

Like ``threading.local`` class, this class supports subclassing, in
which case it will re-execute the ``__init__`` method for each one of
the tracked objects:

.. code:: python

  call_counter = 0

  class CustomLocal(Local):

      a_value = None

      def __init__(self, a_value):
          self.a_value = a_value
          nonlocal call_counter
          call_counter += 1

  mylocal = CustomLocal('foo')

Here ``mylocal.a_value`` will be initialized to ``foo`` for every
tracked object (asyncio's tasks by default). Here ``call_counter``
will count the number of every tracked object in which the ``mylocal``
object has been accessed.

Testing
+++++++

To run the tests you should run the following at the package root::

  python setup.py test

Build status
++++++++++++

.. image:: https://travis-ci.org/azazel75/metapensiero.asyncio.tasklocal.svg?branch=master
    :target: https://travis-ci.org/azazel75/metapensiero.asyncio.tasklocal


CHANGES.rst
.. -*- coding: utf-8 -*-

Changes
-------


0.1 (2015-12-14)
~~~~~~~~~~~~~~~~

- First release complete with tests

0.2 (2015-12-26)
~~~~~~~~~~~~~~~~

- Fix CHANGES and packaging


src/metapensiero/asyncio/tasklocal/tests/test_local.py
# -*- coding: utf-8 -*-
# :Project:  metapensiero.asyncio.tasklocal -- Local class tests
# :Created:    lun 14 dic 2015 12:07:25 CET
# :Author:    Alberto Berti <alberto@metapensiero.it>
# :License:   GNU General Public License version 3 or later
#

import asyncio

import pytest

from metapensiero.asyncio.tasklocal import Local

def test_per_task_var(event_loop):
    mylocal = Local()

    @asyncio.coroutine
    def task_a():
        mylocal.a_var = 'foo'
        yield from asyncio.sleep(0.001) # switches to task_b
        assert mylocal.a_var == 'foo'

    @asyncio.coroutine
    def task_b():
        mylocal.a_var = 'bar'
        yield from asyncio.sleep(0.001) # switches back to task_a
        assert mylocal.a_var == 'bar'

    a = asyncio.Task(task_a(), loop=event_loop)
    b = asyncio.Task(task_b(), loop=event_loop)
    event_loop.run_until_complete(asyncio.wait((a, b), loop=event_loop))
    # check cleanup is working
    # as mylocal has been created, there will be one dict remaining
    # for when current_task() == None
    assert len(mylocal._manager.dicts) == 1

def test_local_subclass(event_loop):

    call_counter = 0

    class CustomLocal(Local):

        a_value = None

        def __init__(self, a_value):
            self.a_value = a_value
            nonlocal call_counter
            call_counter += 1

    mylocal = CustomLocal('foo')

    @asyncio.coroutine
    def task_a():
        nonlocal call_counter
        assert mylocal.a_value == 'foo'
        assert call_counter == 2
        yield from asyncio.sleep(0.001) # switches to task_b
        assert call_counter == 3

    @asyncio.coroutine
    def task_b():
        nonlocal call_counter
        assert mylocal.a_value == 'foo'
        assert call_counter == 3
        yield from asyncio.sleep(0.001) # switches back to task_a
        assert call_counter == 3

    assert call_counter == 1
    assert mylocal.a_value == 'foo'
    a = asyncio.Task(task_a(), loop=event_loop)
    b = asyncio.Task(task_b(), loop=event_loop)
    event_loop.run_until_complete(asyncio.wait((a, b), loop=event_loop))


src/metapensiero/asyncio/tasklocal/tests/test_discriminator.py
# -*- coding: utf-8 -*-
# :Project:  metapensiero.asyncio.tasklocal -- Discriminator tests
# :Created:    dom 13 dic 2015 23:15:17 CET
# :Author:    Alberto Berti <alberto@metapensiero.it>
# :License:   GNU General Public License version 3 or later
#

import asyncio

import pytest

from metapensiero.asyncio.tasklocal import TaskDiscriminator

def test_current(event_loop):

    disc = TaskDiscriminator(loop=event_loop)

    @asyncio.coroutine
    def task_a():
        ctask = asyncio.Task.current_task(loop=event_loop)
        assert ctask is a
        assert disc.current() is ctask
        yield from asyncio.sleep(0.001) # this will force a switch to the
        # other task
        assert disc.current() is ctask

    @asyncio.coroutine
    def task_b():
        ctask = asyncio.Task.current_task(loop=event_loop)
        assert ctask is b
        assert disc.current() is ctask
        yield from  asyncio.sleep(0.001)
        assert disc.current() is ctask

    a = asyncio.Task(task_a(), loop=event_loop)
    b = asyncio.Task(task_b(), loop=event_loop)
    assert a is not b
    event_loop.run_until_complete(asyncio.wait((a, b), loop=event_loop))


def test_dispose(event_loop):

    disc = TaskDiscriminator(loop=event_loop)

    a_completed = False
    b_completed = False

    def a_dispose_handler(_):
        nonlocal a_completed
        a_completed = True

    def b_dispose_handler(_):
        nonlocal b_completed
        b_completed = True

    @asyncio.coroutine
    def task_a():
        disc.dispose(a_dispose_handler)

    @asyncio.coroutine
    def task_b():
        disc.dispose(b_dispose_handler)

    a = asyncio.Task(task_a(), loop=event_loop)
    b = asyncio.Task(task_b(), loop=event_loop)
    event_loop.run_until_complete(asyncio.wait((a, b), loop=event_loop))
    assert a_completed is True
    assert b_completed is True


src/metapensiero/asyncio/tasklocal/__init__.py
# -*- coding: utf-8 -*-
# :Project:   metapensiero.asyncio.tasklocal -- An asyncio's Task-local variable container
# :Created:   dom 09 ago 2015 12:57:35 CEST
# :Author:    Alberto Berti <alberto@metapensiero.it>
# :License:   GNU General Public License version 3 or later
# :Copyright: Copyright (C) 2015 Alberto Berti
#

import asyncio
import weakref
import functools

__all__ = ('BaseDiscriminator', 'TaskDiscriminator', 'Local')


class BaseDiscriminator:
    """The base class for an object devoted to manage a generic set of
    omogenous objects where only one is 'elected' current at a certain
    point in time.
    """

    def current(self):
        """Get the current object"""
        raise NotImplementedError

    def dispose(self, handler):
        """Associate an handler to be called when the object ends its purpose
        or gets finalized.
        """
        obj = self.current()
        if obj:
            weakref.ref(obj, handler)


class TaskDiscriminator(BaseDiscriminator):
    """A specialized discriminator that knows how to deal with tasks"""

    __slots__ = ('loop',)

    def __init__(self, loop=None):
        self.loop = loop

    def current(self):
        return asyncio.Task.current_task(self.loop)

    def dispose(self, handler):
        """The handler gets called when th task completes rather than when it
        is destructed.
        """
        task = self.current()
        if task:
            task.add_done_callback(handler)


class DictManager:
    """A class that manages per-object dict, using the provided
    'discriminator' object or the default that is capable of handling tasks.

    The object returned by discriminator.current() must be hashable.
    """
    __slots__ = ('dicts', 'discriminator')

    def __init__(self, *, discriminator=None):
        self.discriminator = discriminator or TaskDiscriminator()
        self.dicts = {}

    def get_dict(self):
        """Return the dict for the current object  or  raises a KeyError.
        """
        obj_id = hash(self.discriminator.current())
        return self.dicts[obj_id][1]

    def create_dict(self):
        obj_id = hash(self.discriminator.current())
        handler = functools.partial(self._cleanup_dict, obj_id)
        # have to better deal with references?
        self.discriminator.dispose(handler)
        localdict = {}
        # save a ref to the handler in the case the discriminator
        # handles it the weak way
        self.dicts[obj_id] = (handler, localdict)
        return localdict


    def _cleanup_dict(self, obj_id, _):
        if obj_id in self.dicts:
            del self.dicts[obj_id]


class Local:
    """A more generic drop-in replacement for threading.local.

    It doesn't setup any lock so it cannot be considered thread-safe.
    """
    __slots__ = ('_localargs', '_manager', '__dict__')

    def __new__(cls, *args, discriminator=None, **kwargs):
        if (args or kwargs) and (cls.__init__ is object.__init__):
            raise TypeError("Initialization arguments are not supported")
        self = object.__new__(cls)
        manager = DictManager(discriminator=discriminator)
        object.__setattr__(self, '_manager', manager)
        object.__setattr__(self, '_localargs', (args, kwargs))
        manager.create_dict()
        return self

    def _setup_dict(self):
        manager = object.__getattribute__(self, '_manager')
        try:
            map = manager.get_dict()
        except KeyError:
            map = manager.create_dict()
            args, kwargs = object.__getattribute__(self, '_localargs')
            object.__getattribute__(self, '__init__')(*args, **kwargs)
        object.__setattr__(self, '__dict__', map)

    def __getattribute__(self, name):
        object.__getattribute__(self, '_setup_dict')()
        return object.__getattribute__(self, name)

    def __setattr__(self, name, value):
        if name == '__dict__':
            raise AttributeError(
                "%r object attribute '__dict__' is read-only"
                % self.__class__.__name__)
        object.__getattribute__(self, '_setup_dict')()
        return object.__setattr__(self, name, value)

    def __delattr__(self, name):
        if name == '__dict__':
            raise AttributeError(
                "%r object attribute '__dict__' is read-only"
                % self.__class__.__name__)
        object.__getattribute__(self, '_setup_dict')()
        return object.__delattr__(self, name)
